src/transformers/models/mask2former/modeling_mask2former.py

```diff
@@ -0,0 +1,3208 @@
+        hidden_state = input
+        for layer in self.layers:
+            hidden_state = layer(hidden_state)
+        return hidden_state
+
+
+# Copied from transformers.models.maskformer.modeling_maskformer.MaskFormerFPNLayer with MaskFormer->Mask2Former
+class Mask2FormerFPNLayer(nn.Module):
+    def __init__(self, in_features: int, lateral_features: int):
+        """
+        A Feature Pyramid Network Layer (FPN) layer. It creates a feature map by aggregating features from the previous
+        and backbone layer. Due to the spatial mismatch, the tensor coming from the previous layer is upsampled.
+
+        Args:
+            in_features (`int`):
+                The number of input features (channels).
+            lateral_features (`int`):
+                The number of lateral features (channels).
+        """
+        super().__init__()
+        self.proj = nn.Sequential(
+            nn.Conv2d(lateral_features, in_features, kernel_size=1, padding=0, bias=False),
+            nn.GroupNorm(32, in_features),
+        )
+
+        self.block = Mask2FormerFPNConvLayer(in_features, in_features)
+
+    def forward(self, down: Tensor, left: Tensor) -> Tensor:
+        left = self.proj(left)
+        down = nn.functional.interpolate(down, size=left.shape[-2:], mode="nearest")
+        down += left
+        down = self.block(down)
+        return down
+
+
+# Copied from transformers.models.maskformer.modeling_maskformer.MaskFormerFPNModel with MaskFormer->Mask2Former
+class Mask2FormerFPNModel(nn.Module):
+    def __init__(self, in_features: int, lateral_widths: List[int], feature_size: int = 256):
+        """
+        Feature Pyramid Network, given an input tensor and a set of feature map of different feature/spatial size, it
+        creates a list of feature maps with the same feature size.
+
+        Args:
+            in_features (`int`):
+                The number of input features (channels).
+            lateral_widths (`List[int]`):
+                A list with the features (channels) size of each lateral connection.
+            feature_size (int, *optional*, defaults to 256):
+                The features (channels) of the resulting feature maps.
+        """
+        super().__init__()
+        self.stem = Mask2FormerFPNConvLayer(in_features, feature_size)
+        self.layers = nn.Sequential(
+            *[Mask2FormerFPNLayer(feature_size, lateral_width) for lateral_width in lateral_widths[::-1]]
+        )
+
+    def forward(self, features: List[Tensor]) -> List[Tensor]:
+        fpn_features = []
+        last_feature = features[-1]
+        other_features = features[:-1]
+        output = self.stem(last_feature)
+        for layer, left in zip(self.layers, other_features[::-1]):
+            output = layer(output, left)
+            fpn_features.append(output)
+        return fpn_features
+
+def multiscale_deformable_attention_core(value, value_spatial_shapes, sampling_locations, attention_weights):
```

Comment:
A couple of pointers - we use type casting to make sure that the input parameters and output are of the expected type. It is also good practice to use more intuitive names for variable such that users can understand what each line does.

```suggestion
def multiscale_deformable_attention_core(
    value: torch.Tensor,
    value_spatial_shapes: torch.Tensor,
    sampling_locations: torch.Tensor,, 
    attention_weights: torch.FloatTensor,
) -> torch.FloatTensor:
```